src/Client/ClientBase.cpp

```diff
@@ -2136,22 +2137,55 @@ MultiQueryProcessingStage ClientBase::analyzeMultiQueryText(
     }
 
     // INSERT queries may have the inserted data in the query text
-    // that follow the query itself, e.g. "insert into t format CSV 1;2".
-    // They need special handling. First of all, here we find where the
-    // inserted data ends. In multi-query mode, it is delimited by a
-    // newline.
-    // The VALUES format needs even more handling - we also allow the
-    // data to be delimited by semicolon. This case is handled later by
-    // the format parser itself.
-    // We can't do multiline INSERTs with inline data, because most
-    // row input formats (e.g. TSV) can't tell when the input stops,
-    // unlike VALUES.
+    // that follow the query itself, e.g. "insert into t format CSV 1,2".
+    // They need special handling. If Insert format is Values, we use
+    // values format parser to skip valid rows until a single semicolon.
+    // The end of other insert statements is directly determined by \n\n.
     auto * insert_ast = parsed_query->as<ASTInsertQuery>();
     const char * query_to_execute_end = this_query_end;
-
     if (insert_ast && insert_ast->data)
     {
-        this_query_end = find_first_symbols<'\n'>(insert_ast->data, all_queries_end);
+        if (insert_ast->format == "Values")
+        {
+            // try to find the end of INSERT INTO ... VALUES query
+            ReadBufferFromMemory data_in(insert_ast->data, all_queries_end - insert_ast->data);
+            skipBOMIfExists(data_in);
+            for (;;)
```

Comment:
Why not just `while (ValuesBlockInputFormat::skipToNextRow(&data_in, 1, 0));`?

Reply:
Done